okapi-core/src/main/java/org/folio/okapi/util/DepResolution.java

```diff
@@ -788,4 +796,132 @@ static void sortTenantModules(List<TenantModuleDescriptor> tml,
+    Set<String> unknown = new HashSet<>();
+
+    for (Map.Entry<String, ModuleDescriptor> entry : modsEnabled.entrySet()) {
+      ModuleDescriptor md = entry.getValue();
+      boolean optionalUnknown = false;
+      for (InterfaceDescriptor opt : md.getOptionalList()) {
+        if (!DepResolution.findModulesForRequiredInterface(modsEnabled, opt).isEmpty()) {
+          continue; // this optional interface is part of our enabled modules
+        }
+        Map<String, ModuleDescriptor> modules =
+            DepResolution.findModulesForRequiredInterface(modsAvailable, opt);
+        if (modules.isEmpty()) {
+          optionalUnknown = true;
+        } else {
+          getDefinedPermissions(defined, modules.values().iterator().next());
+        }
+      }
+      getReferPermissions(md, referRequired, referModulePermissions, referSubPermissions,
+          optionalUnknown ? unknown : null);
+      getDefinedPermissions(defined, md);
+    }
+    Map<ModuleDescriptor,List<String>> errors = new HashMap<>();
+    for (Map.Entry<String,Set<ModuleDescriptor>> ent : referModulePermissions.entrySet()) {
+      String permissionName = ent.getKey();
+      if (!defined.containsKey(permissionName) && !unknown.contains(permissionName)) {
+        for (ModuleDescriptor md : ent.getValue()) {
+          errors.computeIfAbsent(md, x -> new ArrayList<>()).add("Undefined permission '"
+              + permissionName + "' in modulePermissions");
+        }
+      }
+    }
+    for (Map.Entry<String,Set<ModuleDescriptor>> ent : referRequired.entrySet()) {
+      String permissionName = ent.getKey();
+      if (!defined.containsKey(permissionName) && !unknown.contains(permissionName)) {
+        for (ModuleDescriptor md : ent.getValue()) {
+          errors.computeIfAbsent(md, x -> new ArrayList<>()).add("Undefined permission '"
+              + permissionName + "' in permissionsRequired");
+        }
+      }
+    }
+    for (Map.Entry<String,Set<ModuleDescriptor>> ent : referSubPermissions.entrySet()) {
+      String permissionName = ent.getKey();
+      if (!defined.containsKey(permissionName) && !unknown.contains(permissionName)) {
+        for (ModuleDescriptor md : ent.getValue()) {
+          errors.computeIfAbsent(md, x -> new ArrayList<>()).add("Undefined permission '"
+              + permissionName + "' in subPermissions");
+        }
+      }
+    }
+    for (Map.Entry<String,Set<ModuleDescriptor>> ent: defined.entrySet()) {
+      if (ent.getValue().size() > 1) {
+        String permissionName = ent.getKey();
+        String names = ent.getValue()
+            .stream().map(x -> x.getId()).collect(Collectors.joining(", "));
+        for (ModuleDescriptor md : ent.getValue()) {
+          errors.computeIfAbsent(md, x -> new ArrayList<>()).add("Permission '"
+              + permissionName + "' defined in multiple modules: " + names);
+        }
+      }
+    }
+    return errors;
+  }
+
+  private static void getReferPermissions(
```

Comment:
`addPermissions` might be a better name because it returns void and doesn't get from `refer*` but writes into them.

Reply:
ok